File: lib/features/marketplace/payments/domain/entities/payment_entity.dart
import 'package:math_matric/shared/entities/payment_status.dart';

class PaymentEntity {
  final String id;
  final String bookingId;
  final String studentId;
  final String tutorId;

  final int amountCents;
  final String currency;

  final PaymentStatus status;

  final String? provider;
  final String? providerPaymentId;

  final DateTime createdAt;
  final DateTime updatedAt;

  final DateTime? paidAt;
  final String? failureReason;

  const PaymentEntity({
    required this.id,
    required this.bookingId,
    required this.studentId,
    required this.tutorId,
    required this.amountCents,
    required this.currency,
    required this.status,
    required this.provider,
    required this.providerPaymentId,
    required this.createdAt,
    required this.updatedAt,
    required this.paidAt,
    required this.failureReason,
  });
}

File: functions/src/payments/payment/payment_entity.ts
export const PaymentStatus = {
  pending: "pending",
  processing: "processing",
  paid: "paid",
  failed: "failed",
  cancelled: "cancelled",
  refunded: "refunded",
  partiallyRefunded: "partially_refunded",
} as const;

export type PaymentStatus = typeof PaymentStatus[keyof typeof PaymentStatus];

export interface Payment {
  id: string;
  bookingId: string;
  studentId: string;
  tutorId: string;

  amountCents: number; //original payment, at all times: refundedAmountCents + refundReservedAmountCents <= amountCents
  refundedAmountCents: number; //money that is successfully refunded.
  refundReservedAmountCents: number; // money currently allocated to pending/processing refunds.
  currency: "ZAR";

  status: PaymentStatus;

  provider: string | null;
  providerPaymentId: string | null;

  createdAt: Date;
  updatedAt: Date;

  paidAt: Date | null;
  failureReason: string | null;
}

// For example:

// Payment: R500

// refundedAmountCents = R0
// refundReservedAmountCents = R0
// available = R500

// Student requests R500 refund:

// refunded = R0
// reserved = R500
// available = R0

// Provider succeeds:

// refunded = R500
// reserved = R0
// available = R0
